### ISSUE TITLE
Unsupervised Machine Translation using Monolingual Corpora Only

### ISSUE THREAD

**Issue by howardyclo** [howardyclo/papernotes#1]
### Metadata
- Authors: Guillaume Lample, Ludovic Denoyer, Marc'Aurelio Ranzato 
- Organization: Facebook AI Research
- Conference: ICLR 2018
- Link: https://openreview.net/forum?id=rkYTTf-AZ

**Comment by howardyclo** [comment]
![Figure 1](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Figure%201.png?raw=true)

### Motivation
- NMT requires large-scale parallel corpora in order to achieve good performance.
- Even translating low resource language pairs requires tens of thousands of data.
- Monolingual data is much easier to find.

---

### Model
- Use a single encoder and a single decoder for both domains.
- Use sequence-to-sequence model with attention (Bahdanau et al. 2015).

---

### Training Objectives
#### Initialization
The model starts with an unsupervised naive translation model obtained by making **word-by-word translation** of sentences using a parallel dictionary learned in an unsupervised way (Conneau et al. 2017). Then, at each iteration, the model are trained by minimizing an objective function that measures their ability to both **reconstruct** and **translate** from a **noisy** input sentence.

#### Denoising Auto-Encoding (Critical to performance)
![Equation 1](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Equation%201.png?raw=true)
Δ is the sum of token-level cross-entropy loss between the sentence *x* and the reconstruction from the noisy sentence *x_hat*. The noisy model *C* is created by dropping and swapping tokens in the sentence.

#### Cross Domain Training (Back-translation)
![Equation 2](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Equation%202.png?raw=true)
Learn to reconstruct the sentence *x* from *C(y)*, where *y = M(x)* is the translation of *x*, *M* is the current translation model, and Δ is again the sum of token-level cross-entropy loss.

#### Adversarial Training
Jointly train a discriminator to classify the language given the encoding of source sentences and the encoding of target sentences (Ganin et al. 2016). In detail, the discriminator operates on a sequence of encoded hidden state vectors, and produces a binary prediction (0: source; 1: target). The discriminator is trained by minimizing the cross-entropy loss:

<p align="center">
<img width="50%" src="https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Equation%203-1.png?raw=true" alt="Equation 3-1"></img>
</p>

The encoder is trained instead to fool the discriminator: 

![Equation 3-2](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Equation%203-2.png?raw=true)

#### Final Objective
![Equation 4](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Equation%204.png?raw=true)
λ*auto*, λ*cd*, λ*adv* are hyper-parameters.

---

### Training Algorithm
![Algorithm 1](https://github.com/howardyclo/papernotes/blob/master/images/Unsupervised%20Machine%20Translation%20using%20Monolingual%20Corpora%20Only/Algorithm%201.png?raw=true)

---

### References
- [Improving Neural Machine Translation Models with Monolingual Data](https://arxiv.org/abs/1511.06709) by Sennrich et al., 2015.
- [Word Translation without Parallel Data](https://arxiv.org/pdf/1710.04087.pdf) by Conneau et al., 2017.
- [Neural Machine Translation by Jointly Learning to Align and Translate](https://arxiv.org/abs/1409.0473) by Bahdanau et al., 2015.
- [Domain-Adversarial Training of Neural Networks](https://arxiv.org/abs/1505.07818) by Ganin et al., 2016.

**Comment by howardyclo** [comment]
For comparison to the other similar paper: [Unsupervised Neural Machine Translation](https://arxiv.org/pdf/1710.11041.pdf) by Artetex et al. 2017, please refer to the [slides](https://docs.google.com/presentation/d/12yOLpTD7iiNuotJypvpEVvwRTHIDo-t80TXV0othp58/edit?usp=sharing).

**Comment by elyarAbad** [comment]
Hi!
Sorry! In the published paper, it is mentioned that "We will release the code to the public once the revision process is over". I will be really THANKFUL if you share the link to source code.
Thank you very much!
